README
Vyatta configuration wrapper for System Security Services Daemon


debian/changelog
vyatta-sssd (0.55) unstable; urgency=medium

  [ Nicholas Brown ]
  * dh-systemd is part of debhelper

 -- Daniel Gollub <dgollub@att.com>  Tue, 16 Mar 2021 15:06:12 +0100

vyatta-sssd (0.54) unstable; urgency=medium

  [ Duncan Eastoe ]
  * Add 10-sssd cli-sandbox post create hook script
  * Only mount privileged PAM pipe in sandboxes

 -- Daniel Gollub <dgollub@att.com>  Tue, 23 Feb 2021 21:29:18 +0100

vyatta-sssd (0.53) unstable; urgency=medium

  * SSSD: Configure sssd to run as unprivileged user
  * SSSD.py: Configure sssd to only start nss service
  * SSSD.py: configure sssd process to switch user

 -- Duncan Eastoe <duncan.eastoe@att.com>  Fri, 16 Oct 2020 16:56:04 +0100

vyatta-sssd (0.52) unstable; urgency=medium

  * vyatta_update_service_users: Use consistent indent
  * vyatta_update_service_users: Use universal_newlines
  * Local.pm: Use FQ name when updating user password
  * Local.pm: Handle new database location

 -- Duncan Eastoe <duncan.eastoe@att.com>  Tue, 08 Sep 2020 16:48:54 +0100

vyatta-sssd (0.51) unstable; urgency=medium

  * Remove vyatta-sssd-cfg.init script
  * Migrate to Python 3
  * SSSD: Remove use of try_set_domain_option()
  * Disable NSS local_negative_timeout
  * SSSD: switch tacplus dom to offline_timeout_max
  * SSSD: propagate errors from tacplus dom providers
  * vyatta-sssd-tacacs: Make pam_mkhomedir.so optional

 -- Duncan Eastoe <duncan.eastoe@att.com>  Tue, 28 Jul 2020 11:35:02 +0100

vyatta-sssd (0.50) unstable; urgency=medium

  * Remove vyatta-system-login-sssd-noyang package
  * control: Fix "Priority: extra" Lintian warning
  * Remove debian/lintian
  * Cleanup packaging
  * control: Add missing substvars
  * debian: Run `wrap-and-sort -s`
  * Add tech support script

 -- Duncan Eastoe <duncan.eastoe@att.com>  Fri, 03 Apr 2020 10:28:45 +0100

vyatta-sssd (0.49) unstable; urgency=medium

  [ Mike Manning ]
  * vyatta-sssd: code injection due to Local.pm module

  [ Duncan Eastoe ]
  * Set filter_users for NSS service in sssd.conf

 -- Duncan Eastoe <duncan.eastoe@att.com>  Mon, 10 Feb 2020 14:59:48 +0000

vyatta-sssd (0.48) unstable; urgency=medium

  * DANOS Import master

 -- Vyatta Package Maintainers <DL-vyatta-help@att.com>  Fri, 08 Nov 2019 16:48:57 +0000


tests/resources_service_users_local.sh
function create_user() {
	set resources service-users local user $1 auth plaintext-password "foo"
	set resources service-users local user $1 full-name "Dummy $1"
}

function setup_full_testbed() {
	set resources service-users local group sales-dep
	set resources service-users local group hr-dep

	create_user foobar
	set resources service-users local user foobar group sales-dep
	set resources service-users local user foobar group hr-dep

	create_user jdoe
	set resources service-users local user jdoe group hr-dep
}

function lookup_user() {
	ldbsearch -H /var/lib/sss/db/sssd.ldb -b cn=users,cn=LOCAL,cn=sysdb "(name=$1)" 2> /dev/null | grep -q "0 entries" && return 1
        return 0
}

function lookup_group() {
	ldbsearch -H /var/lib/sss/db/sssd.ldb -b cn=groups,cn=LOCAL,cn=sysdb "(name=$1)" 2> /dev/null | grep -q "0 entries" && return 1
        return 0
}


function reset_sssd_nextid() {

	( cat << EOF
dn: cn=LOCAL,cn=sysdb
changetype: modify
replace: nextID
nextID: 500
EOF
	) | ldbmodify -H /var/lib/sss/db/sssd.ldb -i &> /dev/null
}


(
sss_userdel foobar
sss_userdel jdoe
sss_groupdel sales-dep
sss_groupdel hr-dep
sss_groupdel marketing-dep
sss_groupdel it-dep

for testuser in $( ldbsearch -H /var/lib/sss/db/sssd.ldb -b cn=users,cn=LOCAL,cn=sysdb name | egrep "^name:" | sed 's/name: //g' ); do
	sss_userdel $testuser;
done

) &> /dev/null


echo "Cleanup"
delete resources service-users local &> /dev/null
commit
reset_sssd_nextid



echo "Setup testbed"
setup_full_testbed
commit

echo "Testcase: verify correct SSSD user creation"
id foobar@LOCAL &> /dev/null || echo FAILED_SSSD_CREATE_USER1
lookup_user foobar || echo FAILED_SSSD_SYSDB_CREATE_USER1
id jdoe@LOCAL &> /dev/null || echo FAILED_SSSD_CREATE_USER2
sss_groupshow vyatta-service-users 2> /dev/null | grep -q foobar || echo FAILED_CRATE_USER1_SERVICE_GROUP_ASSIGNMENT
sss_groupshow vyatta-service-users 2> /dev/null | grep -q jdoe || echo FAILED_CRATE_USER2_SERVICE_GROUP_ASSIGNMENT

echo "Testcase: verify correct SSSD group assignment"
groups foobar@LOCAL | grep -q hr-dep || echo FAILED_SSSD_USER_GROUP
groups foobar@LOCAL | grep -q sales-dep || echo FAILED_SSSD_USER_GROUP2

groups jdoe@LOCAL | grep -q hr-dep || echo FAILED_SSSD_USER2_GROUP
groups jdoe@LOCAL | grep -q sales-dep && echo FAILED_SSSD_USER2_GROUP2


show resources service-users local user foobar group | grep -q sales-dep || echo FAILED_USER_GROUP
show resources service-users local user foobar group | grep -q hr-dep || echo FAILED_USER_GROUP2
show resources service-users local user jdoe group | grep -q sales-dep && echo FAILED_WRONG_USER_GROUP
show resources service-users local user foobar full-name | grep -q "Dummy foobar" || echo FAILED_USER_FULL_NAME
show resources service-users local user foobar auth plaintext-password | grep -q "" || echo FAILED_ONLY_ENCRYPTED_PASSWORD
show resources service-users local user foobar auth encrypted-password | grep -q " \$6" || echo FAILED_ENCRYPTED_PASSWORD

echo "Testcase: service-user change password"
set resources service-users local user foobar auth plaintext-password "foobar"
commit
show resources service-users local user foobar auth plaintext-password | grep -q "" || echo FAILED_ONLY_ENCRYPTED_CHANGED_PASSWORD
show resources service-users local user foobar auth encrypted-password | grep -q " \$6" || echo FAILED_ENCRYPTED_CHANGED_PASSWORD

echo "Testcase: service-user change full-name"
set resources service-users local user foobar full-name "Foo Bar"
commit
show resources service-users local user foobar full-name | grep -q "Foo Bar" || echo FAILED_CHANGING_FULLNAME



echo "Testcase: locking of local service-user"
set resources service-users local user foobar lock
commit
show resources service-users local user foobar | egrep -q lock || echo FAILED_USER_LOCK_USER1
ldbsearch -H /var/lib/sss/db/sssd.ldb -b cn=users,cn=LOCAL,cn=sysdb "(name=foobar)" disabled 2> /dev/null | grep -q "disabled: true"  || echo FAILED_USER_LOCK_SYSDB_USER1

echo "Testcase: unlocking of local service-user"
delete resource service-users local user foobar lock
commit
show resources service-users local user foobar | egrep -q lock && echo FAILED_USER_UNLOCK_USER1
ldbsearch -H /var/lib/sss/db/sssd.ldb -b cn=users,cn=LOCAL,cn=sysdb "(name=foobar)" disabled 2> /dev/null | grep -q "disabled: true"  && echo FAILED_USER_UNLOCK_SYSDB_USER1

echo "Testcase: check for stale group ownership in CLI, when group is deleted"
delete resources service-users local group sales-dep
commit

show resources service-users local user foobar group | grep -q sals-dep && echo FAILED_USER_STALE_GROUP
show resources service-users local user foobar group | grep -q hr-dep || echo FAILED_USER_GROUP_ON_CLEANUP__VYATTA_CFG_KNOWN_ISSUE


### This case is not valid, due to FAILED_USER_GROUP_ON_CLEANUP__VYATTA_CFG_KNOWN_ISSUE
### hr-dep group alrady got deleted unintetionally in previous testcase
#echo "Testcase: delete group membership of a user. Expected: group still exists, user is no longer group membership"
#delete resources service-users local user foobar group hr-dep
#commit
#
#id foobar@LOCAL &> /dev/null || echo FAILED_GROUP_USER1_LOST
#groups foobar@LOCAL | grep -q hr-dep && echo FAILED_TO_DROP_HR_GROUP_SSSD
#show resources service-users local user foobar | grep -q hr-dep && echo FAILED_TO_DROP_HR_GROUP_CLI

echo "Testcase: delete entrie group section with multiple group entries"
set resources service-users local group marketing-dep
set resources service-users local group it-dep
commit
delete resources service-users local group
commit
show resources service-users local | grep -q group && echo FAILED_DELETE_ENTRIE_GROUP

echo "Testcase: delete single service-user"
delete resources service-users local user foobar
commit
id foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER1
lookup_user foobar && echo FAILED_SSSD_SYSDB_DELETE_USER1
groups foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP
show resources service-users local user | grep -q foobar && echo FAILLED_DELETE_USER1

echo "Testcase: delete entire service-user user section with multiple user entries"
create_user foobar
commit
delete resources service-users local user
commit
id foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_USER_SECTION_USER1
lookup_user foobar && echo FAILED_SSSD_SYSDB_DELETE_ENTIRE_USER_SECTION_USER1
id jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_USER_SECTION_USER2
groups foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_USER_SECTION_USER1
groups jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_USER_SECTION_USER2
show resources service-users local | grep -q user && echo FAILLED_DELETE_USER_SECTION

echo "Testcase: delete multiple users and groups by deleting entire local service-users section"
setup_full_testbed
commit
id foobar@LOCAL &> /dev/null || echo FAILED_ROUND2_SSSD_CREATE_USER1
lookup_user foobar || echo FAILED_ROUND2_SSSD_SYSDB_CREATE_USER1
id jdoe@LOCAL &> /dev/null || echo FAILED_ROUND2_SSSD_CREATE_USER2
delete resources service-users local
commit
show resources service-users | grep -q local && echo FAILLED_DELETE_LOCAL_SECTION
lookup_user foobar && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_USER1
id foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_LOCAL_SECTION_USER1
groups foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_LOCAL_SECTION_USER1
lookup_group foobar && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_USER1_GROUP

id jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_LOCAL_SECTION_USER2
groups jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_LOCAL_SECTION_USER2

lookup_group sales-dep && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_GROUP1
lookup_group hr-dep && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_GROUP2

echo "Testcase: delete multiple users and groups by deleting entire service-users section"
setup_full_testbed
commit
id foobar@LOCAL &> /dev/null || echo FAILED_ROUND3_SSSD_CREATE_USER1
id jdoe@LOCAL &> /dev/null || echo FAILED_ROUND3_SSSD_CREATE_USER2
delete resources service-users
commit
show resources | grep -q service-users && echo FAILLED_DELETE_SERVICE_USERS_SECTION
id foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_SERVICE_USERS_SECTION_USER1
lookup_user foobar && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_USER1

groups foobar@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_SERVICE_USERS_SECTION_USER1
lookup_group foobar && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_USER1_GROUP

id jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_ENTIRE_SERVICE_USERS_SECTION_USER2
groups jdoe@LOCAL &> /dev/null && echo FAILED_SSSD_DELETE_USER_GROUP_ENTIRE_SERVICE_USERS_SECTION_USER2

lookup_group sales-dep && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_GROUP1
lookup_group hr-dep && echo FAILED_SSSD_DELETE_SYSDB_ENTIRE_LOCAL_SECTION_GROUP2


PAM_SERVICE_USERS_CONF="/etc/pam.d/vyatta-service-users.conf"
echo "Testcase: verify generated service-user PAM configuration file"
setup_full_testbed
commit
test -f $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_NOT_GENERATED_MISSING
grep -q "^auth required pam_deny.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_AUTH_DENY
grep -q "^account required pam_deny.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_ACCOUNT_DENY
egrep -q "^auth sufficient pam_sss.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_AUTH_PAM_SSS
egrep -q "^account sufficient pam_sss.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_ACCOUNT_PAM_SSS
egrep -q "^auth sufficient pam_sss.so domains=local" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_INVALID_AUTH_PAM_SSS_CONFIG
egrep -q "^account sufficient pam_sss.so domains=local" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_INVALID_ACCOUNT_PAM_SSS_CONFIG

echo "Testcase: verify locked service-user PAM configuration file with SSSD disabled"
delete resource service-users
commit
test -f $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_NOT_GENERATED_MISSING
grep -q "^auth required pam_deny.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_AUTH_DENY
grep -q "^account required pam_deny.so" $PAM_SERVICE_USERS_CONF || echo FAILED_PAM_CONF_MISSING_ACCOUNT_DENY
egrep -q "^auth (required|sufficient) pam_sss.so" $PAM_SERVICE_USERS_CONF && echo FAILED_PAM_CONF_NOT_EXPECTING_AUTH_PAM_SSS
egrep -q "^account (required|sufficient) pam_sss.so" $PAM_SERVICE_USERS_CONF && echo FAILED_PAM_CONF_NOT_EXPECTING_ACCOUNT_PAM_SSS






function bench_batch_user_creation() {

	NUMBER_OF_USERS=$1;
	echo "Testcase: creating $NUMBER_OF_USERS service-users in one commit"
	reset_sssd_nextid
	START_SET=$( date +%s )
	for n in $(seq 1 $NUMBER_OF_USERS); do
	     create_user testuser$n
	done
	START_COMMIT=$( date +%s )
	commit
	END_COMMIT=$( date +%s )
	echo "SET time: $(( $START_COMMIT - $START_SET ))s"
	echo "COMMIT time: $(( $END_COMMIT - $START_COMMIT ))s"
	delete resources service-users
	reset_sssd_nextid
	commit
}

bench_batch_user_creation 100
#bench_batch_user_creation 200
#bench_batch_user_creation 300
#bench_batch_user_creation 400
#bench_batch_user_creation 450
bench_batch_user_creation 500



tests/resources_service_users_auth_ldap.sh
set resources service-users ldap acme.com base-dn ou=People,dc=acme,dc=com
set resources service-users ldap acme.com search-filter uid=ldapuser1
set resources service-users ldap acme.com tls cacert /config/ssl/ldap-ca.pem
set resources service-users ldap acme.com tls reqcert allow
set resources service-users ldap acme.com url ldap://192.168.122.161/
set interfaces openvpn vtun1 auth ldap acme.com

set resources service-users ldap fab.acme.com base-dn ou=Machine,dc=acme,dc=com
set resources service-users ldap fab.acme.com search-filter uid=machineuser1
set resources service-users ldap fab.acme.com tls cacert /config/ssl/ldap-ca.pem
set resources service-users ldap fab.acme.com tls reqcert allow
set resources service-users ldap fab.acme.com url ldap://192.168.122.161/
set interfaces openvpn vtun1 auth ldap fab.acme.com


lib/Vyatta/SSSD/PAM.pm
# Copyright (c) 2019 AT&T intellectual property.
# All rights reserved.
#
# Copyright (c) 2014, 2017 Brocade Communications Systems, Inc.
# All rights reserved.
#
# SPDX-License-Identifier: LGPL-2.1-only

package Vyatta::SSSD::PAM;

use warnings;
use strict;

use lib "/opt/vyatta/share/perl5";

use Vyatta::Config;
use Vyatta::SSSD::Local;
use Authen::PAM qw( :constants );

use Exporter q{import};
our @EXPORT = qw( setup_pam_for_service pam_validate_user );

my $SERVICE_PREFIX = "vyatta-";


sub setup_local_domain {
    my ($config_space, $service_instance) = @_;

    my $config = new Vyatta::Config;

    # For each service instance, one internal group gets created.
    # This internal group name must start with vyatta- and will be not
    # disabled inside the CLI.
    my $service_group = $service_instance;
    add_group($service_group);

    # Handle service-user users
    $config->setLevel($config_space." local user");
    foreach my $user ( $config->returnValues() ) {
       user_add_group($user, $service_group);
    }

    foreach my $user ( local_users() ) {
       next if $config->exists("$user");
       user_remove_group($user, $service_group);
    }

    # Handle service-user groups
    $config->setLevel($config_space." local group");
    foreach my $group_member ( $config->returnValues() ) {
       groupmod_append_group($group_member, $service_group);
    }

    foreach my $group_member ( list_sssd_group_member_groups($service_group) ) {
       next if $config->exists("$group_member");
       groupmod_remove_group($group_member, $service_group);
    }
}

sub generate_pam {
    my ($config_space, $service_instance) = @_;

    my $service_user_domains = undef;
    my $pam_sss_service_users = "";
    my $pam_sss_service_users_local = "";

    my $config = new Vyatta::Config;
    $config->setLevel($config_space);
    my @auth_methods = $config->listNodes();

    foreach my $auth_method ( @auth_methods ) {
         next if ($auth_method eq 'local');

         foreach my $domain ( $config->returnValues($auth_method) ) {
            if (defined($service_user_domains)) {
               $service_user_domains = $service_user_domains . ',';
            }
            $service_user_domains = $service_user_domains . $domain;
         }
    }


    if (defined($service_user_domains)) {
       $pam_sss_service_users = "
auth sufficient pam_sss.so domains=$service_user_domains
account sufficient pam_sss.so domains=$service_user_domains
";
    }

    if (grep { $_ eq 'local' } @auth_methods) {
       $pam_sss_service_users_local = "
auth sufficient pam_sss.so domains=local
account required pam_sss.so domains=local
account sufficient pam_succeed_if.so user ingroup $service_instance
";
    }

    my $content = "##
## This file is auto-generated by Vyatta:SSSD:PAM.pm.
## Do not edit, all changes will be lost!
##

$pam_sss_service_users
$pam_sss_service_users_local

auth requisite pam_deny.so
account requisite pam_deny.so
";

    my $path = "/etc/pam.d/".$service_instance.".conf";
    open(my $fd, '>', $path)
         or die "Can't generate authentication configuration! Can't open file $path: $!";
    print $fd $content;
    close $fd;
}

sub setup_pam_for_service {
    my ($config_space, $service, $instance) = @_;

    my $service_instance = $SERVICE_PREFIX . $service ."-". $instance;

    my $config = new Vyatta::Config;
    $config->setLevel($config_space);
    my @auth_methods = $config->listNodes();

    setup_local_domain($config_space, $service_instance)
       if (grep { $_ eq 'local' } @auth_methods);

    generate_pam($config_space, $service_instance);
}

#
# pam_validate_user verifies via PAM if a user exists in a configured PAM service
# configuration. This allows to check if service PAM configuration would grant a
# user XYZ access, by checking LDAP domains, local domain or other SSSD domains which
# are configured in that specific PAM service configuration.
#
# Known issues:
#  - LDAP domains: search-filter gets ignored. Only base-dn gets matchted. This results
#                  in more matches then actually wanted. (E.g. users which would be
#                  excluded by the search-filter)
# - LOCAL domains: pam accgt_mgmt require pam conversation / authentication and fails with
#                  PAM_AUTH_ERR. This is still good enough to indicate that this user
#                  would be granted access to the particular service interface.
#
sub pam_validate_user {
        my ($service, $instance, $user) = @_;

        die "User is required for PAM validation" unless($user);

        my $pam_service = $SERVICE_PREFIX . $service ."-". $instance .".conf";
        my $pam = Authen::PAM->new( $pam_service, $user );
        die "PAM verification failed while preparing PAM handle" unless($pam);

        return undef unless -e "/etc/pam.d/$pam_service";

        my $result = $pam->pam_acct_mgmt;

	# PAM_AUTH_ERR is also an indication for SSSD and pam_sss that a user
	# exists. PAM_AUTH_ERR get returned if a local domain users exists.
	# The PAM_AUTH_ERR is due to authentication attempts with pam conversation,
	# which fails since no pam conversion handler is set.
        return 1 if ($result == PAM_SUCCESS || $result == PAM_AUTH_ERR);
        return undef;
}

1;


Vyatta/SSSD.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2019-2020 AT&T intellectual property.
# All rights reserved.
#
# Copyright (c) 2014, 2017 by Brocade Communications Systems, Inc.
# All rights reserved.
#
# SPDX-License-Identifier: LGPL-2.1-only

import os
import syslog
import SSSDConfig

class SSSD(SSSDConfig.SSSDConfig):

	SCHEMA_API='/usr/share/doc/sssd/sssd.api.d'
	SCHEMA_FILE='/usr/share/doc/sssd/sssd.api.conf'
	SSSD_CONFIG = '/etc/sssd/sssd.conf'
	SSSD_USER = 'sssd'

	# Not yet requested sssd-ldap option are disabled for now
        # until requested.
	LDAP_OPTS = {
		'url': 'ldap_uri',
		'base-dn': 'ldap_search_base',
		'schema': 'ldap_schema', # rfc2307 vs. rfc2307bis
		'bind-dn': 'ldap_default_bind_dn',
		'password': 'ldap_default_authtok',

		## User attributes
		#'user search-base': 'ldap_user_search_base',
		#'user object-class': 'ldap_user_object_class',
		#'user name-attribute': 'ldap_user_name',
		#'user uid-attribute': 'ldap_user_uid_number',
		#'user gid-attribute': 'ldap_user_gid_number',
		#'user gecos-attribute': 'ldap_user_gecos',
		#'user home-directory-attribute': 'ldap_user_home_directory',
		#'user shell-attribute': 'ldap_user_shell',
		#'user uuid-attribute': 'ldap_user_uuid',
		#'user cn-attribute': 'ldap_user_fullname',
		#'user member-of-attribute': 'ldap_user_member_of',

		#'force-upper-case-realm': 'ldap_force_upper_case_realm',

		## Group
		'group base-dn': 'ldap_group_search_base',
		#'group object-class': 'ldap_group_object_class',
		#'group name-attribute': 'ldap_group_name',
		#'group gid-attribute': 'ldap_group_gid_number',
		'group member-attribute': 'ldap_group_member',
		#'group uuid': 'ldap_group_uuid',

		## TLS
		'tls reqcert': 'ldap_tls_reqcert',
		'tls cacert': 'ldap_tls_cacert',
		#'tls cacertdir': 'ldap_tls_cacertdir', # would required maintaince of hashed cert dir
		#'tls id-use-start-tls': 'ldap_id_use_start_tls', # force TLS also for ID lookup, not only auth

		## SASL not yet supported
		#'sasl mech': 'ldap_sasl_mech',
		#'sasl authid': 'ldap_sasl_authid',

		## Misc
		#'pwd-policy': 'ldap_pwd_policy', # password expiration policy on client side
		#'dns-service-name': 'ldap_dns_service_name', # DNS-SD not yet supported

		'follow-referrals': 'ldap_referrals',
		'search-filter': 'ldap_access_filter',

                ## KRB not yet supported
		#'krb5 keytab': 'ldap_krb5_keytab',
		#'krb5 init-creds': 'ldap_krb5_init_creds',
		#'krb5 realm': 'krb5_realm',

		#'user-principal': 'ldap_user_principal', # krb specific

		#'default-authtok-type': 'ldap_default_authtok_type', # only one value possible: password

		## Timeout settings

		# might change in future in SSSD, with more fine-grain settings:
		#'timeout search': 'ldap_search_timeout',

		#'timeout network': 'ldap_network_timeout', # default: 5 seconds
		#'timeout opt': 'ldap_opt_timeout',         # default: 5 seconds
        }

	TACPLUS_OPTS = {
		# internal only: 'tacplus_shell',
		# internal only: 'tacplus_service',
		# internal only: 'tacplus_proto',
		# internal only: 'tacplus_user_gid',
		# internal only: 'tacplus_homedir',
	}

	def __init__(self):

		if not os.path.exists(self.SCHEMA_API):
			self.SCHEMA_API = None

		if not os.path.exists(self.SCHEMA_FILE):
			self.SCHEMA_FILE = None

		SSSDConfig.SSSDConfig.__init__(self, schemaplugindir=self.SCHEMA_API, schemafile=self.SCHEMA_FILE)

		if os.path.exists(self.SSSD_CONFIG):
			self.import_config(self.SSSD_CONFIG)
		else:
			self.new_config()

		# Always perform setup of the SSSD and NSS services
		self._setup_sssd()
		self._setup_nss()

	def try_set_domain_option(self, domain, opt, value):
		try:
			domain.set_option(opt, value)
		except SSSDConfig.NoOptionError as e:
			syslog.syslog(syslog.LOG_DEBUG, "Error setting option '{}' " \
				"on domain '{}': {}".format(opt, domain.get_name(), e))

	def setup_local(self):
		try:
			local_domain = self.get_domain('local')
		except SSSDConfig.NoDomainError:
			local_domain = self.new_domain('local')

		local_domain.add_provider('local', 'id')
		local_domain.add_provider('local', 'auth')
		# local domain does not provide an access provider
		#local_domain.add_provider('local', 'access')

		local_domain.set_option('enumerate', 'true')
		local_domain.set_option('min_id', 500)
		# No upper limit, required since there is no ID rotation:
		# http://www.redhat.com/archives/freeipa-devel/2009-March/msg00076.html
		# max_id is then UINT32_MAX
		local_domain.set_option('max_id', 0)
		local_domain.set_option('default_shell', '/bin/false')

		local_domain.set_active(True)
		self.save_domain(local_domain)

	def setup_ldap(self, name, options):
		try:
			ldap_domain = self.get_domain(name)
		except SSSDConfig.NoDomainError:
			ldap_domain = self.new_domain(name)

		ldap_domain.add_provider('ldap', 'id')
		ldap_domain.add_provider('ldap', 'auth')
		ldap_domain.add_provider('ldap', 'access')
		ldap_domain.set_active(True)

		# Following options are by default in SSSD true, but in the CLI
		# those are CLI:
		#
		#  * ldap_referrals
		#
                # Since we are using typeless-boolean, we disable those first. And
		# enable them if those are set in the CLI.

		for opt in ['ldap_referrals']:
			ldap_domain.set_option(opt, 'false')

		option_schema = ldap_domain.list_options()
		for opt in options:
			if option_schema[opt][0] == bool:
				ldap_domain.set_option(opt, 'true')
			else:
				ldap_domain.set_option(opt, options[opt])

		self.save_domain(ldap_domain)

	def setup_tacplus(self, name, options):
		try:
			tacplus_domain = self.get_domain(name)
		except SSSDConfig.NoDomainError:
			tacplus_domain = self.new_domain(name)

		tacplus_domain.add_provider('tacplus', 'id')
		tacplus_domain.add_provider('tacplus', 'auth')
		#tacplus_domain.add_provider('tacplus', 'access')
		tacplus_domain.set_active(True)

		option_schema = tacplus_domain.list_options()
		for opt in options:
			if option_schema[opt][0] == bool:
				tacplus_domain.set_option(opt, 'true')
			else:
				tacplus_domain.set_option(opt, options[opt])

		tacplus_domain.set_option('tacplus_shell', '/bin/vbash')
		tacplus_domain.set_option('tacplus_service', 'vyatta-exec')
		tacplus_domain.set_option('tacplus_secrets', 'true')
		tacplus_domain.set_option('tacplus_proto', 'login')
		tacplus_domain.set_option('tacplus_user_gid', '100')
		tacplus_domain.set_option('tacplus_homedir', '/var/tmp/aaa-home/%u')
		tacplus_domain.set_option('entry_cache_user_timeout', '3600')
		tacplus_domain.set_option('min_id', '2000')
		tacplus_domain.set_option('offline_timeout', '60')
		tacplus_domain.set_option('offline_timeout_max', '0')
		tacplus_domain.set_option('propagate_provider_error', 'true')


		self.save_domain(tacplus_domain)

	def _setup_sssd(self):
		# Setup the [sssd] section
		try:
			sssd_serv = self.get_service("sssd")
		except SSSDConfig.NoServiceError:
			sssd_serv = self.new_service("sssd")

		sssd_serv.set_option("user", self.SSSD_USER)
		sssd_serv.set_option("switch_user", 'true')
		sssd_serv.set_option("services", ['nss'])
		self.save_service(sssd_serv)

	def _setup_nss(self):
		try:
			nss_serv = self.get_service("nss")
		except SSSDConfig.NoServiceError:
			nss_serv = self.new_service("nss")

		# NSS requests for filtered users are ignored.
		# "*" is ignored since a request for this user is generated
		# when path completion is requested in Bash, due to a bug in
		# the completion script. See Debian bug #825317.
		# root is ignored by default so is also set here.
		nss_serv.set_option("filter_users", "root,*")

		# By default local users are kept in the negative cache for
		# 4 hours. When there is a negative cache hit the data provider
		# (eg. TACACS+) is not queried, which means we don't determine
		# whether the provider is offline, which results in local user
		# fallback not working reliably.
		# Setting it to 0 means that all negative cache entries timeout
		# according to entry_negative_timeout, which defaults to 15 seconds.
		nss_serv.set_option("local_negative_timeout", 0)

		self.save_service(nss_serv)


	def commit(self):
		# Created files needs to be owned by root and root group.
                # If not SSSD refuses to start. Even the file SSSD creates during
                # startup require to have the gid() set.
		prevGid = os.getgid()
		os.setgid(0)

		self.write(self.SSSD_CONFIG)

		# Generate /etc/default/sssd
		etc_default_sssd_content = """##
## This file is auto-generated by vyatta-sssd.
## Do not edit, all changes will be lost!
##
"""

		if len(self.list_domains()) == 0:
			# SSSD fails to start on boot-up with no domains
			# configured. Due to absence of chkconfig and friends
			# SSSD init-startup get avoided by the sourced default sssd
			# config file, which does not exist by default.
			etc_default_sssd_content += """
# No SSS domains configured - disabling SSSD -vyatta-sssd
exit 0
"""

		with open('/etc/default/sssd', 'w') as f:
			f.write(etc_default_sssd_content)

		# Once the default/sssd configuration got written we invoke
		# the SSSD service.
		if len(self.list_domains()) > 0:
			os.system('systemctl restart sssd &> /dev/null');
		else:
			os.system('systemctl stop sssd &> /dev/null');

		os.setgid(prevGid)

